File: demo-android/app/src/test/java/com/example/demo/pact/TasksCreatePactTest.kt
package com.example.demo.pact

import au.com.dius.pact.consumer.MockServer
import au.com.dius.pact.consumer.dsl.PactBuilder
import au.com.dius.pact.consumer.junit5.PactConsumerTestExt
import au.com.dius.pact.consumer.junit5.PactTestFor
import au.com.dius.pact.core.model.V4Pact
import au.com.dius.pact.core.model.annotations.Pact
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(PactConsumerTestExt::class)
@PactTestFor(providerName = "demo-service-tasks-create")
class TasksCreatePactTest {

    @Pact(consumer = PactFixtures.CONSUMER)
    fun createTaskSuccessPact(builder: PactBuilder): V4Pact = builder.legacyHttpPact {
        given("task title is unique")
            .uponReceiving("a valid task creation request")
            .path("/v1/tasks")
            .method("POST")
            .headers("Content-Type", "application/json")
            .body(PactFixtures.createTaskRequestBody())
            .willRespondWith()
            .status(201)
            .headers(mapOf("Content-Type" to "application/json"))
            .body(PactFixtures.taskResponseBody())
    }

    @Test
    @PactTestFor(pactMethod = "createTaskSuccessPact")
    fun shouldHaveCreateTaskContractWhenPostingValidTask(mockServer: MockServer) {
        runBlocking {
            // Given
            val api = PactFixtures.createTaskApi(mockServer)

            // When
            api.createTask(PactFixtures.createTaskRequest())
        }
    }
}

File: demo-react-native/src/test/pact/tasks.create.pact.test.ts
import {MatchersV3} from '@pact-foundation/pact';

import {ErrorResponse, Task, TaskPriority, TaskRequest, TaskStatus} from '@/data/models/task';
import {createTaskApi} from '@/data/remote/taskApi';
import {createPact, likeTask, TASK_ID} from './tasks.pact.fixtures';

const { regex, fromProviderState } = MatchersV3;

const pact = createPact('demo-service-tasks-create');

const task: TaskRequest = {
  title: 'Prepare release notes',
  description: 'Document release tasks',
  status: TaskStatus.TODO,
  priority: TaskPriority.MEDIUM,
};

const requestBody: { [K in keyof TaskRequest]: unknown } = {
  title: fromProviderState('${taskTitle}', task.title),
  description: task.description,
  status: task.status,
  priority: task.priority,
};

describe('tasks POST /v1/tasks pact', () => {
  it('should have create task success contract when posting valid task', async () => {
    const createdTask: Task = {
      id: TASK_ID,
      title: task.title,
      description: task.description,
      status: task.status,
      priority: task.priority,
      createdDate: '2026-04-26T09:00:00.000Z',
      updatedDate: '2026-04-26T09:00:00.000Z',
    };

    await pact
      .addInteraction()
      .given('task title is unique')
      .uponReceiving('a valid task creation request')
      .withRequest('POST', '/v1/tasks', (req) => {
        req.headers({ 'Content-Type': 'application/json' });
        req.jsonBody(requestBody);
      })
      .willRespondWith(201, (res) => {
        res.headers({ 'Content-Type': 'application/json' });
        res.jsonBody(likeTask(createdTask));
      })
      .executeTest(async (mockServer) => {
        const api = createTaskApi(`${mockServer.url}/v1/`);
        await api.createTask(task);
      });
  });

  it('should have create task duplicate contract when posting duplicate title', async () => {
    const errorBody: { [K in keyof Required<ErrorResponse>]: unknown } = {
      message: regex(
        "^Task with title '.*' already exists$",
        `Task with title '${task.title}' already exists`,
      ),
    };

    await pact
      .addInteraction()
      .given('task title already exists')
      .uponReceiving('a task creation request with duplicate title')
      .withRequest('POST', '/v1/tasks', (req) => {
        req.headers({ 'Content-Type': 'application/json' });
        req.jsonBody(requestBody);
      })
      .willRespondWith(409, (res) => {
        res.headers({ 'Content-Type': 'application/json' });
        res.jsonBody(errorBody);
      })
      .executeTest(async (mockServer) => {
        const api = createTaskApi(`${mockServer.url}/v1/`);
        await expect(api.createTask(task)).rejects.toThrow('already exists');
      });
  });
});
